// viem/_cjs/tempo/Addresses.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.zoneOutbox = exports.validator = exports.tip403Registry = exports.tip20Factory = exports.stablecoinDex = exports.signatureVerifier = exports.receivePolicyGuard = exports.pathUsd = exports.nonceManager = exports.feeManager = exports.addressRegistry = exports.accountRegistrar = exports.accountKeychain = exports.accountImplementation = void 0;
exports.accountImplementation = '0x7702c00000000000000000000000000000000000';
exports.accountKeychain = '0xaAAAaaAA00000000000000000000000000000000';
exports.accountRegistrar = '0x7702ac0000000000000000000000000000000000';
exports.addressRegistry = '0xfdc0000000000000000000000000000000000000';
exports.feeManager = '0xfeec000000000000000000000000000000000000';
exports.nonceManager = '0x4e4F4E4345000000000000000000000000000000';
exports.pathUsd = '0x20c0000000000000000000000000000000000000';
exports.receivePolicyGuard = '0xB10C000000000000000000000000000000000000';
exports.signatureVerifier = '0x5165300000000000000000000000000000000000';
exports.stablecoinDex = '0xdec0000000000000000000000000000000000000';
exports.tip20Factory = '0x20fc000000000000000000000000000000000000';
exports.tip403Registry = '0x403c000000000000000000000000000000000000';
exports.validator = '0xcccccccc00000000000000000000000000000000';
exports.zoneOutbox = '0x1c00000000000000000000000000000000000002';
