// viem/_cjs/tempo/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebCryptoP256 = exports.WebAuthnP256 = exports.withRelay = exports.withFeePayer = exports.walletNamespaceCompat = exports.Transport = exports.Transaction = exports.TokenIds = exports.Storage = exports.P256 = exports.KeyAuthorizationManager = exports.Hardfork = exports.Formatters = exports.Expiry = exports.tempoActions = exports.Chain = exports.Capabilities = exports.Actions = exports.Addresses = exports.Account = exports.Abis = exports.VirtualMaster = exports.VirtualAddress = exports.TokenId = exports.Tick = exports.TempoAddress = exports.ReceivePolicyReceipt = exports.Period = exports.MultisigConfig = exports.Channel = exports.Secp256k1 = exports.PublicKey = exports.Bytes = void 0;
var ox_1 = require("ox");
Object.defineProperty(exports, "Bytes", { enumerable: true, get: function () { return ox_1.Bytes; } });
Object.defineProperty(exports, "PublicKey", { enumerable: true, get: function () { return ox_1.PublicKey; } });
Object.defineProperty(exports, "Secp256k1", { enumerable: true, get: function () { return ox_1.Secp256k1; } });
var tempo_1 = require("ox/tempo");
Object.defineProperty(exports, "Channel", { enumerable: true, get: function () { return tempo_1.Channel; } });
Object.defineProperty(exports, "MultisigConfig", { enumerable: true, get: function () { return tempo_1.MultisigConfig; } });
Object.defineProperty(exports, "Period", { enumerable: true, get: function () { return tempo_1.Period; } });
Object.defineProperty(exports, "ReceivePolicyReceipt", { enumerable: true, get: function () { return tempo_1.ReceivePolicyReceipt; } });
Object.defineProperty(exports, "TempoAddress", { enumerable: true, get: function () { return tempo_1.TempoAddress; } });
Object.defineProperty(exports, "Tick", { enumerable: true, get: function () { return tempo_1.Tick; } });
Object.defineProperty(exports, "TokenId", { enumerable: true, get: function () { return tempo_1.TokenId; } });
Object.defineProperty(exports, "VirtualAddress", { enumerable: true, get: function () { return tempo_1.VirtualAddress; } });
Object.defineProperty(exports, "VirtualMaster", { enumerable: true, get: function () { return tempo_1.VirtualMaster; } });
exports.Abis = require("./Abis.js");
exports.Account = require("./Account.js");
exports.Addresses = require("./Addresses.js");
exports.Actions = require("./actions/index.js");
exports.Capabilities = require("./Capabilities.js");
exports.Chain = require("./Chain.js");
var Decorator_js_1 = require("./Decorator.js");
Object.defineProperty(exports, "tempoActions", { enumerable: true, get: function () { return Decorator_js_1.decorator; } });
exports.Expiry = require("./Expiry.js");
__exportStar(require("./errors.js"), exports);
exports.Formatters = require("./Formatters.js");
exports.Hardfork = require("./Hardfork.js");
exports.KeyAuthorizationManager = require("./KeyAuthorizationManager.js");
exports.P256 = require("./P256.js");
exports.Storage = require("./Storage.js");
exports.TokenIds = require("./TokenIds.js");
exports.Transaction = require("./Transaction.js");
exports.Transport = require("./Transport.js");
var Transport_js_1 = require("./Transport.js");
Object.defineProperty(exports, "walletNamespaceCompat", { enumerable: true, get: function () { return Transport_js_1.walletNamespaceCompat; } });
Object.defineProperty(exports, "withFeePayer", { enumerable: true, get: function () { return Transport_js_1.withFeePayer; } });
Object.defineProperty(exports, "withRelay", { enumerable: true, get: function () { return Transport_js_1.withRelay; } });
exports.WebAuthnP256 = require("./WebAuthnP256.js");
exports.WebCryptoP256 = require("./WebCryptoP256.js");
