// viem/_cjs/tempo/zones/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.zoneModerato = exports.zone = exports.portalAddresses = exports.getPortalAddress = exports.from = exports.http = exports.Abis = void 0;
exports.Abis = require("./Abis.js");
var transport_js_1 = require("./transport.js");
Object.defineProperty(exports, "http", { enumerable: true, get: function () { return transport_js_1.http; } });
var zone_js_1 = require("./zone.js");
Object.defineProperty(exports, "from", { enumerable: true, get: function () { return zone_js_1.from; } });
Object.defineProperty(exports, "getPortalAddress", { enumerable: true, get: function () { return zone_js_1.getPortalAddress; } });
Object.defineProperty(exports, "portalAddresses", { enumerable: true, get: function () { return zone_js_1.portalAddresses; } });
Object.defineProperty(exports, "zone", { enumerable: true, get: function () { return zone_js_1.zone; } });
Object.defineProperty(exports, "zoneModerato", { enumerable: true, get: function () { return zone_js_1.zoneModerato; } });
